app: Creates winget manifest

vyncent-t opened the pull request:
For issue #1265 

## Description

This PR addresses the request to support `winget`, the built-in package manager for Windows, especially since it's natively available in Windows 11. The goal is to streamline the installation experience for Windows users of the Headlamp desktop app.

## Changes

1. **Winget Manifest**: Introduced a new `winget` manifest folder tailored for the Headlamp desktop app. This manifest will allow users to install Headlamp via the `winget install headlamp` command.

2. Followed the official `winget` documentation to ensure the manifest adheres to recommended guidelines and best practices.

3. Updated installation documentation (if applicable) to include instructions for the `winget` installation method.

## Testing

- [x] Validate the `winget` manifest using the official `winget` CLI tools.

## Notes

illume commented:
The docs for ["Create your package manifest"](https://learn.microsoft.com/en-us/windows/package-manager/package/manifest?tabs=minschema%2Cversion-example) contains docs for the fields.

vyncent-t commented:
I would like to mention a few fields that need some input on ways to be revised. ref file Headlamp.Headlamp.yaml

The single file manifest does validate with the fields currently written, open to input for additional metadata to include.

- PackageIdentifier
Required field, winget asks this to be the name of the app that matched the downloaded App info in windows add or remove program settings. Currently, it is set to `PackageIdentifier: Headlamp.Headlamp` as the convention for this follows `Publisher.App` but the app info after download is still Kinvolk, also not sure if to brand it under Microsoft or CNCF. 
![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/2b64b259-cdba-4442-ac24-4f31e9a857c2)

- Publisher
Required field, will update this field once the discussion is made for the Publisher name choice.

knrt10 commented on app/windows/winget/0.19.0/headlamp.headlamp.installer.yaml:
```diff
@@ -0,0 +1,53 @@
+      - UnsupportedArgument:    	# Optional unsupported Windows Package Manager Client argument
+    ReleaseDate: 2023-03-08                	# Optional release date
+ManifestType: installer
+ManifestVersion: 1.5.0
```
Can you please end new line at the end.

joaquimrocha reviewed (changes requested):
Finally got to review this.
I think if the goal is to submit it to https://github.com/microsoft/winget-pkgs , then we should make these files work as templates, with placeholders like `__PKG_VERSION__` instead of a hardcoded version. And then we have script to fill the details like we have in `choco-bump.sh` and hopefully we can submit a PR every time Headlamp has a release.

Does this make sense?

vyncent-t commented:
This push changes the Headlamp.Headlamp.yaml file from singleton to version, also updates the triple file format for manifest

vyncent-t commented:
This push changes the Headlamp.Headlamp.yaml file from singleton to version, also updates the triple file format for manifest

joaquimrocha reviewed (changes requested):
This looks much cleaner now. Left a suggestion of a new tag.
This however needs to be turned into a template file somehow and we need a script to populate the template and generate the files (which should look like what you wrote in the end). See for example:
https://github.com/headlamp-k8s/headlamp/blob/main/app/windows/chocolatey/choco-bump.sh

With this, we can fork the manifests Github repo and create a PR over to the original repo every time we have a release.

vyncent-t commented:
The manifest was able to validate as a single file, currently working on fixing errors for validation of the new structure

error reads

Manifest validation failed.
Manifest Error: Schema Error. [Schema validation failed.
Error context: <root>[Installers][0][Platform] Description: Value type not permitted by 'type' constraint.
Error context: <root>[Installers][0] Description: Failed to validate against schema associated with property name 'Platform'.
Error context: <root>[Installers] Description: Failed to validate item #0 in array.
Error context: <root> Description: Failed to validate against schema associated with property name 'Installers'.
] File: Headlamp.Headlamp.installer.yaml
Manifest Error: Failed to process field. [Platform]

vyncent-t commented:
Removed the optional platform field from the installer file, this field would prevent the manifest from validating.

Manifest now validates fine

illume commented on app/windows/winget/0.19.1/Headlamp.Headlamp.installer.yaml:
```diff
@@ -0,0 +1,23 @@
+        InstallerType: exe
+    ReleaseDate: 2023-08-24
+ManifestType: installer
+ManifestVersion: 1.5.0
```
Does this file and the others need a new line at the end? I would guess so.

illume reviewed (commented):
Could you please write the command you used to validate it in the PR description somewhere?

Left a note about the missing new lines that github is warning about at the end of the files.

Note there is one conversation left unresolved. Could you please answer that conversation and close it if you have resolved it?

joaquimrocha commented on app/windows/winget/manifests/0.19.1/Headlamp.Headlamp.installer.yaml:
This is not needed in this repo.

joaquimrocha commented on app/windows/winget/manifests/0.19.1/Headlamp.Headlamp.locale.en-US.yaml:
This is not needed in this repo.

joaquimrocha commented on app/windows/winget/winget_create.sh:
```diff
@@ -0,0 +1,55 @@
+
+# Check if the correct number of arguments has been provided
+if [ "$#" -ne 6 ]; then
+    echo "Usage: $0 PACKAGE_VERSION INSTALLER_URL INSTALLER_SHA256 DISPLAY_VERSION RELEASE_DATE RELEASE_NOTES_URL"
```
All of these can be determined from the Headlamp windows exe, except for the release date.

vyncent-t commented:
- removed non generated manifest
- created js script for winget and removed sh script

Note: works similar to the old sh script, need to run the file using `node winget_create.js <version> <sha256>` as params

joaquimrocha commented on app/windows/winget/winget_create.js:
```diff
@@ -0,0 +1,114 @@
+  process.exit(1);
+}
+
+const projectVersion = args[0];
```
We should check if the given version has a `v` prefix or not, and adapt it accordingly or tell the user whether a v is needed.

joaquimrocha commented on app/windows/winget/winget_create.js:
```diff
@@ -0,0 +1,114 @@
+ * 3. Install the node-fetch module with the command `npm install node-fetch`.
+ * 4. Run this script using the command `node winget_create.js <projectVersion> <sha256>`.
+ *
+ * Parameters:
```
We need a destination directory parameter, so we write the new files into that.
I imagine we'll be using it like:
node winget_create.js 0.19.1 /my/path/to/the/wingetrepo/h/headlamp/0.19.1

The given directory should be created if it doesn't exist.

joaquimrocha reviewed (changes requested):
I think it's almost there, but it's missing something we had talked about.

joaquimrocha commented on app/windows/winget/winget_create.js:
```diff
@@ -0,0 +1,144 @@
+
+const args = process.argv.slice(2);
+
+if (args.length > 2) {
```
What you want to check here if whether we have 2 args. I ran the script without arguments for getting some help and I get an exception.

vyncent-t replied:
fixed!

joaquimrocha commented on app/windows/winget/winget_create.js:
```diff
@@ -0,0 +1,144 @@
+
+  // create manifest version directory if it doesnt exist
+  if (!fs.existsSync(manifestDir)) {
+    fs.mkdirSync(manifestDir);
```
Use the recursive option of this function, as it creates the missing directory paths for the user.
We should inform the user as well that we've created the directory, to inform of ongoing changes.

vyncent-t replied:
added and also placed in a try catch to handle errors on creation

joaquimrocha commented on app/windows/winget/winget_create.js:
```diff
@@ -0,0 +1,144 @@
+}
+
+async function updateReleaseInfo(response, tag) {
+  const assets = response.assets;
+
+  const checksums = assets.find(v => v.name === 'checksums.txt');
+
+  const checksumsUrl = checksums.browser_download_url;
```
Remove the empty newlines please.

vyncent-t replied:
Done!

joaquimrocha commented on app/windows/winget/winget_create.js:
```diff
@@ -0,0 +1,144 @@
+      continue;
+    }
+
+    if (filename.includes(`.exe`)) {
```
This should not be a string literal (just a plane string).
It will also break if we ever add another arch to the windows installers. So maybe check for the full suffix which will guarantee we don't break: `x64.exe`.
Even better if you just declare the arch as a global constant and then use it where needed, so it's easy to change the script in the future.

vyncent-t replied:
added!

joaquimrocha commented on app/windows/winget/winget_create.js:
```diff
@@ -0,0 +1,144 @@
+    asset => asset.name === `Headlamp-${tag}-win-x64.exe`
+  ).browser_download_url;
+
+  const releaseDateInfo = response.published_at.slice(0, 10);
```
I am not sure the hardcoded length is robust enough.
Maybe it is, but this line should get a comment explaining why we slice.

vyncent-t replied:
Removed the slice so it will print the fill ISO formatted timestamp

joaquimrocha commented on app/windows/winget/winget_create.js:
Please rename the file to `winget-create.js` in order to match the name format we've been using in e.g. app/scripts .

vyncent-t replied:
Done!

joaquimrocha commented on app/windows/winget/templates/Headlamp.Headlamp.locale.en-US.yaml:
```diff
@@ -0,0 +1,27 @@
+License: Apache 2.0
+LicenseUrl: https://github.com/headlamp-k8s/headlamp/blob/main/LICENSE
+ShortDescription: Kubernetes web UI.
+Description: Headlamp was created to be a Kubernetes web UI that has the traditional functionality of other web UIs/dashboards available (i.e. to list and view resources) as well as other features.
```
I know this was here before. I just wonder now if we should have something clearer, like we have in Github: `A Kubernetes web UI that is fully-featured, user-friendly and extensible`

vyncent-t replied:
Added!

joaquimrocha reviewed (approved):
Looking good! I have left a few more comments mainly so the script is more robust.

joaquimrocha reviewed (changes requested):
Mistakenly had marked it as approved.

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,153 @@
+
+if (!(args.length === 2)) {
+  console.error('Usage: node winget-create.mjs <projectVersion> <directoryPath>');
+  console.error('Please provide the project version without the "v" prefix.');
```
I think this error message is misplaced here.

vyncent-t replied:
Ah! I also copy pasted the top usage command but it was also misspelled for the .js file!

I tried to add more messages to provide more context as to not be too short. I felt the one line of just the usage message didn't explain enough about the correct way to use the command (it would be annoying for someone to have to "unlock" the hidden error comment about dropping the v when they could be told within their first mistake to be sure to include it)

I can remove it if its not practical!

vyncent-t replied:
Removed!

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,153 @@
+}
+
+const projectVersion = args[0];
+
```
Remove empty line.

vyncent-t replied:
Done!

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,153 @@
+let sha256;
+
+const arch = {
+  x64: 'x64',
+  x86: 'x86',
```
We only have one arch: 'x64'. You can keep it in an obj or array but let's not add more arches we don't deal with now.

vyncent-t replied:
Ah okay, I tried to include it to match the original scripts object but where it could be used in the checksum piece!
Removed it!
![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/26670769-dcad-4f28-8df5-2fb2bd876d1c)

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,153 @@
+    'Headlamp.Headlamp.yaml',
+  ];
+
+  // use path to get the template directory
```
This comment is not adding much.

vyncent-t replied:
I will remove all the notes I left in while learning how to make the script work from the original, they're basically to do notes anyway.

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,153 @@
+  // use path to get the template directory
+  const templateDir = path.join(__dirname, 'templates/');
+
+  const manifestDir = path.join(directoryPath, '/', version, '/');
```
You don't need '/' with path.join

vyncent-t replied:
Okay, wasn't sure if I needed to keep them in or not, I had ran into a pathing issue while implementing the request for a custom file path parameter and it seemed to have been able to work once I put them both in

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,153 @@
+  try {
+    // create manifest version directory if it doesnt exist
+    if (!fs.existsSync(manifestDir)) {
+      console.log(`Created directory ${manifestDir}`);
+      fs.mkdirSync(manifestDir, { recursive: true });
```
These lines should be inverted: if mkdirSync fails, then we'd not see "Created directory". Otherwise we may misguide the user.

vyncent-t replied:
Apologies, I had left it in to see what it would print while the mkdir line is commented out!

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,153 @@
+
+  templates.forEach(template => {
+    // read the template file
+    const templateContent = fs.readFileSync(`${templateDir}${template}`, 'utf8');
```
Ideally you'd join the templateDir and template with path.join .

vyncent-t replied:
ah okay, it seemed logical to use temps when using the manual '/' in path technique from earlier!

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,153 @@
+      .replace('__RELEASE_NOTES_URL__', notesURL);
+
+    // write the new content to the manifest directory
+    fs.writeFileSync(`${manifestDir}${template}`, newContent);
```
Ditto for path.join here.

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,153 @@
+      continue;
+    }
+
+    if (filename.includes(arch.x64)) {
+      checksumMap[arch.x64] = checksum;
+    }
```
There's a potential bug here. This code doesn't check whether you have a windows installer here, so if another platform was sorted after the Windows installer in the checksums.txt file, you'd get that checksum here instead of the window's file one.

I suggest that instead of gathering all this info into the map, you first get the file name as you do below and then use it when looking for the checksum. This way you guarantee you really get the checksum for the file you want.

vyncent-t replied:
Done!

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,153 @@
+  const checksumsUrl = checksums.browser_download_url;
+  const checksumMap = await getChecksums(checksumsUrl);
+  const browserDownloadUrlInfo = assets.find(
+    asset => asset.name === `Headlamp-${tag}-win-x64.exe`
```
This generated file name can be used for a `getChecksum(checksumsURL, fileName)` function.

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,153 @@
+  const browserDownloadUrlInfo = assets.find(
+    asset => asset.name === `Headlamp-${tag}-win-x64.exe`
+  ).browser_download_url;
+  const releaseDateInfo = response.published_at;
```
So now you no longer slice the date info. How do you get it without the time part?

vyncent-t replied:
I thought by the original manual slice not being robust enough you meant it cut out too much information so I left it as the full ISO timestamp but I can add the slice back in if you no longer want the full timestamp or I can use split I think

vyncent-t replied:
Added slice back in and const for context

joaquimrocha reviewed (changes requested):
Found some more issues.

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,151 @@
+      console.log(`Created directory ${manifestDir}`);
+    }
+  } catch (err) {
+    console.error(`Error creating directory ${manifestDir}: ${err}`);
```
There should be a `process.exit(1)` if something went wrong with creating the directory.

vyncent-t replied:
done!

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,151 @@
+  const assets = response.assets;
+  const checksums = assets.find(v => v.name === 'checksums.txt');
+  const checksumsUrl = checksums.browser_download_url;
+  const browserDownloadName = assets.find(
+    asset => asset.name === `Headlamp-${tag}-win-x64.exe`
+  ).name;
+  const browserDownloadUrlInfo = assets.find(
+    asset => asset.name === `Headlamp-${tag}-win-x64.exe`
```
This processes the assets list twice. Please get store the resolved asset into an `asset` variable. So you don't have to repeat the logic.

vyncent-t replied:
done!

illume commented:
This needs the DCO fixed... `git commit --amend -s`

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,149 @@
+    const templateContent = fs.readFileSync(templatePath, 'utf8');
+
+    const newContent = templateContent
+      .replace(/__PACKAGE_VERSION__/g, version)
```
This format is inconsistent with the rest. Can you make it `'__PACKAGE_VERSION__'`.

joaquimrocha commented on app/windows/winget/winget-create.js:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    if (filename === fileName) {
+      checksumMap[arch.x64] = checksum;
```
There's no point in submitting a map when we only have one value.

vyncent-t replied:
done!

joaquimrocha reviewed (changes requested):
I have reviewed it again. It's almost there so I will take it from here and update the script.

joaquimrocha reviewed (approved):
It LGTM now and it works great!
